contract/lst-oracle/src/lib.rs

```diff
@@ -0,0 +1,189 @@
+    ) -> bool {
+        result.unwrap_or(false)
+    }
+
+    pub fn list_ema_prices_no_older_than(
+        &self,
+        price_ids: Vec<PriceIdentifier>,
+        age: u64,
+    ) -> PromiseOrValue<OracleResponse> {
+        if price_ids.is_empty() {
+            return PromiseOrValue::Value(OracleResponse::new());
+        }
+
+        let (dispatched_price_ids, promises): (Vec<_>, Vec<_>) = price_ids
+            .iter()
+            .copied()
+            .map(|price_id| {
+                if let Some(entry) = self.transformers.get(&price_id) {
+                    (entry.price_id, Some(entry.call.promise()))
+                } else {
+                    (price_id, None)
+                }
+            })
+            .unzip();
+
+        let mut promise = ext_pyth::ext(self.oracle_id.clone())
+            .list_ema_prices_no_older_than(dispatched_price_ids, age);
+
+        for p in promises.into_iter().flatten() {
+            promise = promise.and(p);
+        }
+
+        PromiseOrValue::Promise(
+            promise.then(
+                self_ext!(Gas::from_tgas(4))
+                    .list_ema_prices_no_older_than_01_consume_results(price_ids),
+            ),
+        )
+    }
+
+    #[private]
+    pub fn list_ema_prices_no_older_than_01_consume_results(
+        &self,
+        original_price_ids: Vec<PriceIdentifier>,
+    ) -> OracleResponse {
+        fn callback_result<T: DeserializeOwned>(index: u64) -> T {
+            match env::promise_result(index) {
+                PromiseResult::Successful(vec) => {
+                    serde_json::from_slice(&vec).unwrap_or_else(|e| env::panic_str(&e.to_string()))
+                }
+                PromiseResult::Failed => env::panic_str(&format!("Promise index {index} failed")),
+            }
+        }
+
+        let oracle_result = callback_result::<OracleResponse>(0);
+        let mut result = OracleResponse::with_capacity(oracle_result.len());
+
+        let mut i = 1;
+        for price_id in original_price_ids {
+            if let Some(price) = oracle_result.get(&price_id) {
+                result.insert(price_id, price.clone());
+            } else {
+                let Some(entry) = self.transformers.get(&price_id) else {
+                    env::panic_str(&format!(
+                        "No transformer associated with price ID: {price_id}",
+                    ));
+                };
+                let Some(price) = oracle_result.get(&entry.price_id) else {
+                    env::panic_str(&format!(
+                        "Mapped price ID is not in oracle result: {price_id}",
+                    ));
+                };
+                let input = callback_result::<Decimal>(i);
+                i += 1;
+
+                result.insert(
+                    price_id,
+                    price
+                        .clone()
+                        .and_then(|price| entry.action.apply(price, input)),
+                );
+            }
+        }
```

Comment:
Changed to use built in `iter` and enumerate, `i + 1` because `i` starts at 0 with enumerate.

```suggestion
        for (i, price_id) in original_price_ids.iter().enumerate() {
            if let Some(price) = oracle_result.get(&price_id) {
                result.insert(price_id, price.clone());
            } else {
                let Some(entry) = self.transformers.get(&price_id) else {
                    env::panic_str(&format!(
                        "No transformer associated with price ID: {price_id}",
                    ));
                };
                let Some(price) = oracle_result.get(&entry.price_id) else {
                    env::panic_str(&format!(
                        "Mapped price ID is not in oracle result: {price_id}",
                    ));
                };
                let input = callback_result::<Decimal>(i + 1);

                result.insert(
                    price_id,
                    price
                        .clone()
                        .and_then(|price| entry.action.apply(price, input)),
                );
            }
        }
```

Reply:
This is functionally different, since price IDs earlier in the list may not have instigated an additional promise, so it's not always its positional index + 1.

Reply:
Ah - yeah, my mistake. Will change it now.

Reply:
You are right - good point. Disregard this suggestion.